### ISSUE TITLE
False negative for same-named functions from different structs.

### ISSUE THREAD

**Issue by maboesanman** [rust-analyzer/rust-analyzer#7098]
I am using `pin_project` quite a bit, so rust analyzer loses track of the types pretty often, but it is reporting a missing argument because I have two functions with the same name and rust analyzer believes I am using one when I am using the other.

I'm not really sure how to go about making a minimum reproducible example for this, so I am including a link to my repo, with no warnings or compile errors which exhibits the described behavior on macOS 10.15.7 

https://github.com/maboesanman/cozal/tree/c9a732e0fee3985dc4275b760a617017b7c301e5

the `src/core/transposer/transposer_update.rs` file has errors on lines 50, 151, and 155 which seem to be confusing the init function of the `transposer` trait (`src/core/transposer/transposer.rs:59`) for the init functions defined at `src/core/transposer/curried_input_future.rs:50` and `src/core/transposer/curried_schedule_future.rs:65`.

the errors go away if I rename the transposer trait's init function and calls to it to something else.

**Comment by maboesanman** [comment]
additionally, I am using the rust-analyzer vs-code extension

**Comment by lnicola** [comment]
![image](https://user-images.githubusercontent.com/308347/103370818-7ae1b080-4ad6-11eb-8c36-6c71782b8d09.png)

So we don't know what `self.project()` is, `inner` has an unknown type, and we pick the wrong `init` method. @flodiebold were you saying something about removing this fallback?

And `self.project()` probably doesn't work because we don't support attribute proc macros: https://github.com/rust-analyzer/rust-analyzer/issues/6029

**Comment by maboesanman** [comment]
this is my workaround, if it's relevant:

```rust
pub fn init_pinned(self: Pin<&mut Self>) {
    let inner: Pin<&mut TransposerUpdateInner<'a, T>> = self.project().0;
    inner.init();
}
```

**Comment by flodiebold** [comment]
> @flodiebold were you saying something about removing this fallback?

No, but see #6975.

**Comment by skade** [comment]
I have a rather minimal reproduction of this:

```rust
pin_project! {
    pub struct ToProject {
        item: Option<String>
    }
}

pub fn test(mut test: Pin<&mut ToProject>) {
    let proj = test.project();
    let s = proj.item.take().unwrap(); // "Expected 1 argument, found 0" 
}
```

This gives `Expected 1 argument, found 0` for `take()`, as it prefers `Iterator::take` over `Option::take`.
